Reload pipeline descriptions as required

briandealwis opened the pull request:
A configuration tab's `isValid()` may be called even when the tab is not active (i.e., after a `deactivated()`/`performApply()` and — more importantly — before an `activated()`/`initializeFrom()`), and it may be called frequently.  This PR improves loading and detecting changed launch configuration details.

Fixes #2232

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core.test/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/MajorVersionPipelineRunnerConfigurationTest.java:
```diff
@@ -66,4 +72,21 @@ public void testDefaultRunnerInConfiguredRunners() {
+      public Object answer(InvocationOnMock invocation) throws Throwable {
+        // for getAttribute(attr, default)
+        Object[] arguments = invocation.getArguments();
+        return arguments.length == 2 ? arguments[1] : null;
```
See if this works: `when(empty.getAttribute(anyObject(), anyObject())).thenAnswer(AdditionalAnswers.returnsSecondArg());`

briandealwis replied:
Ooo, that's much nicer.

chanseokoh commented:
Let's not close #2232. I know at least another case where #2232 happens, and I verified that this PR doesn't cover that case.

chanseokoh commented:
Oops, I was on a wrong branch. This seems to fix #2232.

chanseokoh commented:
But again, I was able to reproduce #2232 with https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2232#issuecomment-320368325.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineLaunchConfiguration.java:
```diff
@@ -49,26 +48,30 @@
-    PipelineLaunchConfiguration configuration = createDefault();
+      ILaunchConfiguration launchConfiguration, MajorVersion projectVersion) throws CoreException {
+    PipelineLaunchConfiguration configuration =
+        new PipelineLaunchConfiguration(defaultRunner(projectVersion));
```
` ... = createDefault(projectVersion)` looks simpler.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java:
```diff
@@ -262,7 +269,9 @@ public void clear() {
   public void setDefaults(ILaunchConfigurationWorkingCopy configuration) {
-    launchConfiguration = PipelineLaunchConfiguration.createDefault();
+    project = findProject(configuration);
+    MajorVersion version = getMajorVersion(project);
```
Set it to the field `majorVersion`?

briandealwis replied:
Not appropriate here — `setDefaults()` is called when a new launch configuration is created, e.g., using the create-new button from the Run/Debug Configurations dialog.  It's intended to pre-populate the launch config with suitable defaults.

chanseokoh replied:
If so, is it OK that we save `project` just above (and `launchConfiguration` below)? It looks inconsistent that you save `project` and `launchConfiguration` while you don't for `majorVersion`.

briandealwis commented:
> I can still reproduce #2232 with #2232 (comment).

Hmm, me too.

briandealwis commented:
`updateHierarchy()` computes the hierarchy in the background, and isValid() either needs to wait for it *or* we can just return `false` since we're not in a valid state.
```
private void updateHierarchy()
      throws InvocationTargetException, InterruptedException {
    getLaunchConfigurationDialog().run(true, true, new IRunnableWithProgress() {
      @Override
      public void run(IProgressMonitor monitor)
          throws InvocationTargetException, InterruptedException {
        hierarchy = getPipelineOptionsHierarchy(monitor);
      }
    });
  }
```

chanseokoh commented:
I thought we made `updateHierarchy()` synchronous. My cursory testing aligns with it. (For example, if I put `Thread.sleep(10000)` inside `run()`, `updateHierarchy()` returns only after 10 sec. Not clearly documented, but I think `run()` here runs synchronously, even though `fork` is true. Maybe we should just set if `false`. I believe making everything synchronous was the decision about #2289.

Also, I'd like to point out https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2232#issuecomment-320368325 is not because of concurrency. I left a comment back then: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2232#issuecomment-323175689

chanseokoh commented:
Perhaps https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2232#issuecomment-320368325 is a bit different problem, and you might want to tackle it in a separate PR?

elharo commented:
Is this PR still in play or is it now moot?

chanseokoh commented:
I know this is still in play.

briandealwis commented:
PTAL: I've pushed up a substantial revision that addresses the above points.  The crux of these problem is that we were guessing a project's dataflow version and often getting it wrong.

Highlights of the changes:
  - `PipelineArgumentsTab` and `PipelineLaunchConfiguration` no longer record the pipeline runner until one has been selected
  - `PipelineArgumentsTab#setDefaults()` no longer sets default information
  - Added a new method `DataflowDependencyManager#getActualProjectMajorVersion()` as its `getProjectMajorVersion()` treats an absence as all possible versions, whereas we want to know an exact version.
  - Added support for visibly enabling/disabling the subcomponents of the `PipelineArgumentsTab` when provided an invalid project

chanseokoh commented:
I'll look into it, but it's going to take some time.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core.test/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/MajorVersionPipelineRunnerConfigurationTest.java:
```diff
@@ -66,4 +72,15 @@ public void testDefaultRunnerInConfiguredRunners() {
+  public void testFromLaunchConfiguration_defaultRunner() throws CoreException {
+    ILaunchConfiguration empty =
+        mock(ILaunchConfiguration.class, AdditionalAnswers.returnsSecondArg());
+    IProject project = mock(IProject.class);
```
This is unused. Also, could you rename `lc` to `pipelineLaunchConfig`?

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
+
+    @Test
+    public void testReload()
+        throws CoreException, InvocationTargetException, InterruptedException {
```
unnecessary `throws`

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
+    }
+
+    @Test
+    public void testInitializeFrom_accessibleProject() throws CoreException {
```
unnecessary `throws`

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
-    public void testInitializeForm_noExceptionForNonAccessibleProject() throws CoreException {
-      IWorkspaceRoot workspaceRoot = mock(IWorkspaceRoot.class);
-      when(workspaceRoot.getProject(anyString())).thenReturn(mock(IProject.class));
+    public void testInitializeFrom_noExceptionForNonAccessibleProject() throws CoreException {
```
Unnecessary `throws`.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
+
+    @Test
+    public void testIsValid_errorOnNonDataflowProject()
+        throws CoreException, InvocationTargetException, InterruptedException {
```
Unnecessary `throws`.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
+
+    @Test
+    public void testIsValid_errorOnNonProject()
+        throws CoreException, InvocationTargetException, InterruptedException {
```
Unnecessary `throws`.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/preferences/RunOptionsDefaultsComponent.java:
```diff
@@ -432,6 +432,11 @@ public void selectionChanged(SelectionChangedEvent event) {
     stagingLocationInput.addModifyListener(listener);
   }
 
+  public boolean isEnabled() {
```
This is used only in tests. Probably don't have to be `public` but `@VisibleForTesting`.

chanseokoh replied:
Still pending.

chanseokoh replied:
Weird. I can't compile unless this is `public`, but Travis doesn't seem to have any problem. Should we revert in to `public`? `@VisibleForTesting` could remain.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/page/component/TextAndButtonComponent.java:
```diff
@@ -54,6 +53,11 @@ public void setText(String value) {
     text.setText(value);
   }
 
+  public boolean isEnabled() {
```
This is used only in tests. Probably don't have to be `public` but `@VisibleForTesting`.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineLaunchConfiguration.java:
```diff
@@ -122,20 +123,44 @@ public void setUseDefaultLaunchOptions(boolean useDefaultLaunchOptions) {
+    return Objects.equals(majorVersion, other.majorVersion)
+        && Objects.equals(argumentValues, other.argumentValues)
+        && Objects.equals(runner, other.runner)
+        && Objects.equals(userOptionsName, other.userOptionsName);
```
Should we also compare the `useDefaultLaunchOptions` field?

briandealwis replied:
It should.  I *suspect* I didn't include this as held as have another PR (that needs dusting off) that removes this support (from [#2430](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2430#issuecomment-336514168)).

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineLaunchConfiguration.java:
```diff
@@ -122,20 +123,44 @@ public void setUseDefaultLaunchOptions(boolean useDefaultLaunchOptions) {
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(majorVersion, argumentValues, runner, userOptionsName);
```
Should we factor in `useDefaultLaunchOptions`?

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineLaunchConfiguration.java:
```diff
@@ -122,20 +123,44 @@ public void setUseDefaultLaunchOptions(boolean useDefaultLaunchOptions) {
+        .add("argumentValues", argumentValues).toString();
   }
 
   private void setValuesFromLaunchConfiguration(ILaunchConfiguration configuration)
```
Noticed this should just have been a constructor:
```java
  private PipelineLaunchConfiguration(MajorVersion majorVersion, ILaunchConfiguration configuration)
      throws CoreException {
    this.majorVersion = Preconditions.checkNotNull(majorVersion);

    // never set a default runner unless actually specified
```
I was hunting for a while who can call this when, but if we make this a constructor, it becomes obvious. `fromLaunchConfiguration` then could simply be
```java
  public static PipelineLaunchConfiguration fromLaunchConfiguration(MajorVersion majorVersion,
      ILaunchConfiguration launchConfiguration) throws CoreException {
    return new PipelineLaunchConfiguration(majorVersion, launchConfiguration);
  }
```
Maybe we should just ditch the static `fromLaunchConfiguration()`. This another-static-constructor method doesn't help me understand code, IMO.

But then, this refactoring doesn't have to be done in this PR.

chanseokoh replied:
Hmm... maybe the reason for the static method was that the original author didn't like making a constructor throw an exception?

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineLaunchConfiguration.java:
```diff
@@ -146,30 +171,23 @@ private void setValuesFromLaunchConfiguration(ILaunchConfiguration configuration
-        PipelineConfigurationAttr.RUNNER_ARGUMENT.toString(), runner.getRunnerName());
+    if (runner != null) {
+      configuration.setAttribute(PipelineConfigurationAttr.RUNNER_ARGUMENT.toString(),
+          runner.getRunnerName());
```
Note to myself: save a runner in LC after selecting a project, save, close, unset a project, save, close, and see if PLC is constructed without a runner.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/project/DataflowDependencyManager.java:
```diff
@@ -125,7 +125,23 @@ public boolean hasTrackedDataflowDependency(IProject project) {
+  /**
+   * Returns a possible version range for the given project.
+   * 
+   * @throws IllegalStateException if the version range is invalid
```
I think the Javadoc should be something callers can understand. Unclear to me as a caller when a range is invalid and how to avoid it.

chanseokoh replied:
Looking at the code a bit, I think `IllegalStateException` should never be thrown, unless our code is buggy. It's when, e.g., doing `.createFromVersionSpec("nonsensical string")`, but we are sane to always pass a well-formed version string.

briandealwis replied:
This code pulls out the version specified in the user's `pom.xml` for the Dataflow Java SDK (`com.google.cloud.dataflow:google-cloud-dataflow-java-sdk-all`), and it's possible the user has invalid values.  I'll update the javadoc to make this more explicit.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/project/DataflowDependencyManager.java:
```diff
@@ -125,7 +125,23 @@ public boolean hasTrackedDataflowDependency(IProject project) {
+   * Returns the encoded version range for the given project or {@code null} if not found.
+   * 
+   * @return the version range or {@code null} if not found
+   * @throws IllegalStateException if the version range is invalid
```
I think the Javadoc should be something callers can understand. Unclear to me as a caller when a range is invalid and how to avoid it.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
+      dependencyManager = mock(DataflowDependencyManager.class);
+
+      project1 = mockProject("project1");
+      when(workspaceRoot.getProject(eq("project1"))).thenReturn(project1);
```
Nit: you don't need `eq` when all arguments are constants.

briandealwis replied:
`eq()` is necessary if you use pattern matchers, as we do in `testSuppressingUpdates()`

      when(workspaceRoot.getProject(anyString())).thenReturn(mock(IProject.class));

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui.test/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTabTest.java:
```diff
@@ -61,35 +77,150 @@
+      doReturn(MajorVersion.ONE).when(dependencyManager).getProjectMajorVersion(project1);
+
+      project2 = mockProject("project2");
+      when(workspaceRoot.getProject(eq("project2"))).thenReturn(project2);
```
Nit: you don't need `eq` when all arguments are constants.

briandealwis commented:
> @chanseokoh: I was following #2232 (comment), and when I hit Apply after entering a Dataflow project in the Main tab, I got this logged:

Oh geesh.  Ok, here the issue is that the user hasn't actually opened the Pipeline Arguments tab, so no changes have been made and thus there are no changes for this tab to apply.

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java:
```diff
@@ -117,6 +118,15 @@ public void setPreferences(DataflowPreferences preferences) {
     }
   }
 
+  public boolean isEnabled() {
```
Used only in tests. Maybe making it non-public?

chanseokoh commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java:
```diff
@@ -298,50 +312,97 @@ PipelineRunner getSelectedRunner() {
+      DataflowUiPlugin.logError(ex, "Error while initializing from existing configuration"); //$NON-NLS-1$
+      project = null;
+      launchConfiguration = null;
+      return true;
```
Maybe `return false`.

briandealwis replied:
Returning false means the values were not changed.  Added a comment.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java:
```diff
@@ -352,49 +413,40 @@ void updateRunnerButtons(MajorVersion majorVersion) {
-    if (project != null && project.isAccessible()) {
+  private PipelineOptionsHierarchy getPipelineOptionsHierarchy(IProgressMonitor monitor) {
+    if (launchConfiguration != null) {
+      Verify.verify(project != null && project.isAccessible());
```
But if the project is null then the launchConfiguration will be null.

chanseokoh replied:
Ah, I missed that.

chanseokoh commented:
I can still reproduce #2232.

1. Create a Dataflow v1 project and a v2 project.
1. Create a new launch config. Whatever the project value in the _Main_ tab is, set it to the v1 project.
1. Select the _Pipeline Arguments_ tab.
1. Come back to _Main_.
1. Select the v2 project.
1. Select the _Pipeline Arguments_ tab.

briandealwis commented:
This PR as is currently does handle [@chanseokoh's sequence](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2585#issuecomment-358112464) correctly, but in a bit of a round-about fashion.  In verifying it, I realized there's a better way, but it's worth explaining the interactions here.

When I run [@chanseokoh's sequence](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2585#issuecomment-358112464), the launch config dialog (LCD) shows two tabs in error: the _Main_ and _Pipeline Arguments_ tab.  As the _Main_ tab is active, its error is shown: _Main type not specified_).
<img width="496" alt="screen shot 2018-01-17 at 3 35 16 pm" src="https://user-images.githubusercontent.com/202851/35065737-1ec5ae5a-fb9c-11e7-9287-ea95df57be01.PNG">

If I specify a type, then the other error in the _Pipeline Arguments_ tab's error is shown, that an invalid runner is selected:
<img width="539" alt="screen shot 2018-01-17 at 3 37 45 pm" src="https://user-images.githubusercontent.com/202851/35065831-75644848-fb9c-11e7-8e69-131655b740d7.PNG">

Which seems legitimate: a runner had been selected — and encoded into this `ILaunchConfiguration` — but with the project change that runner is not valid for the new project.  When the user selects that tab, the `PipelineArgumentsTab` detects the invalid runner (`PipelineArgumentsTab#updateRunnerButtons()`) and selects a valid runner by _selecting the corresponding runner radio button_ (and logs that it was changed).  This button-selection is important.

Eventually our `initializeFrom()` calls `updateLaunchConfigurationDialog()` (inherited from `AbstractLaunchConfigurationTab`), which initiates a somewhat complicated sequence of actions including calling `performApply(ILaunchConfigurationWorkingCopy)` on the active tab.  `PipelineArgumentTab#performApply()` writes out the selected runner button to the provided `ILaunchConfigurationWorkingCopy`, thus repairing the launch configuration.

I've skipped some details in the previous paragraph: it's actually more complicated.

   - `PipelineArgumentsTab` has modification and selection listeners on all of the UI components in the tab that calls `updateLaunchConfigurationDialog()`
   - `updateLaunchConfigurationDialog()` calls `performApply(ILaunchConfigurationWorkingCopy)` on the active tab and then calls `isValid(ILaunchConfiguration)` on all the tabs to reflect errors
   - Updating the UI components in `initializeFrom(ILaunchConfiguration)` to match the `ILaunchConfig` *currently* triggers these modification listeners and thus results in multiple calls to `updateLaunchConfigurationDialog()`, and thus triggering calls to `performApply()`.
   - Because our actions in `initializeFrom() were calling `updateLaunchConfigurationDialog()` and triggering `performApply()`, **we were writing out stale UI settings to the `ILaunchConfigurationWorkingCopy`**.
   - So I introduced a field `activated` to record whether the tab's UI has been updated to reflect the changes in the _current_ ILaunchConfiguration`; `performApply()` uses this flag to determine whether the UI settings should be reflected into the provided `ILaunchConfigurationWorkingCopy`
      - (I'm going to rename this flag to `uiUpToDate`)
      - As `activated` is only set to `true` at the end of `initializeFrom()`, most of the calls to `performApply()` are (correctly) ignored.  Eventually `activated` is set to `true` and the next `updateLaunchConfigurationDialog()` will trigger the `performApply()` which writes out the changed runner.
   - But `activated` isn't set to `true` until after the UI is up-to-date, and the code to update the pipeline options uses the pipeline settings as pulled in from the `ILaunchConfiguration`, which uses the old runner

### 30,000 ft view

If we step back, activating our `PipelineArgumentsTab` follows three steps:

   1. we load state from an ILaunchConfiguration (`isValid()` and `initializeFrom()`)
   2. we update the UI to reflect that state, including corrections (`initializeFrom()`)
   3. we update the ILaunchConfigurationWorkingCopy from the UI (`performApply()`)

The crux of the problem is that updating the UI in (2) may use invalid state (invalid pipeline runner pulled from the `ILaunchConfiguration`).  And even if we were able to correct the state (i.e., change the pipeline runner in our state — we can't change the ILaunchConfiguration until `performApply()`), updating our UI components triggered calls to `isValid()` which would re-load the state from the `ILaunchConfiguration` and so over-write our state corrections (though not our corrections in the UI).

### The easy fix

It turns out `PipelineArgumentsTab` has support for suppressing these calls to `updateLaunchConfigurationDialog()` from our modification/selection listeners.  So the easy fix is to

   1. suppress these calls during the UI updating in `initializeFrom()`
   2. correct the state when detected in `updateRunnerButtons()`
   3. ensure a call to `updateLaunchConfigurationDialog()` happens after updating the UI so as to trigger a `performApply()` to push the corrections into the `ILaunchConfigurationWorkingCopy`.

I've implemented this now and @chanseokoh's sequence now runs with no warnings. \o/

### The better fix

These problems are really because our `PipelineArgumentsTab` serves two purposes: to validate the settings recorded in an `ILaunchConfiguration`, and to drive a UI to reflect and modify the settings recorded in an `ILaunchConfiguration`.  We're using fields to hold onto the state derived from the `ILaunchConfiguration`, and these fields are being modified between calls to `initializeFrom()` and `isValid()`, and so we need to coordinate whether these fields are stale or not.  

We could avoid having this coordination problem by moving the local state from being fields and instead pass it around as method arguments. But we're actually maintaining a lot of information, and as some of it's expensive to compute we'll need to cache.

elharo commented:
If this is ready to go, I say let's take the easy fix. We can file a tech debt bug to improve it, but it may not be worth burning time on.

briandealwis commented:
Weird, I see it too though I'm sure it didn't show up previously.

briandealwis commented:
@chanseokoh found an error:
> go to the Pipleline tab, makes some changes to be able to Apply there. Then selecting a different runner does not enable Apply again, i.e., the tab doesn't notice there is a change.

Cause is that I removed the call to `updateLaunchConfigurationDialog()` from `updatePipelineOptionsForm()` and explicitly called it from `initializeFrom()`, but missed that there were two other locations that called `updatePipelineOptionsForm()`.
